packages/core/src/services/loopDetectionService.ts

```diff
@@ -432,44 +439,123 @@ export class LoopDetectionService {
         'unproductive_state_confidence',
       ],
     };
-    let result;
+
+    const flashResult = await this.queryLoopDetectionModel(
+      'loop-detection',
+      contents,
+      schema,
+      signal,
+    );
+
+    if (
+      flashResult &&
+      typeof flashResult['unproductive_state_confidence'] === 'number'
+    ) {
+      if (
+        flashResult['unproductive_state_confidence'] >= LLM_CONFIDENCE_THRESHOLD
+      ) {
+        if (this.config.isInFallbackMode()) {
+          this.handleConfirmedLoop(flashResult);
+          return true;
+        }
+
+        // Double check with configured model
+        const mainModelResult = await this.queryLoopDetectionModel(
+          'loop-detection-double-check',
+          contents,
+          schema,
+          signal,
+        );
+
+        logLlmLoopCheck(
+          this.config,
+          new LlmLoopCheckEvent(
+            this.promptId,
+            flashResult['unproductive_state_confidence'] as number,
+            this.config.getModel(),
+            typeof mainModelResult?.['unproductive_state_confidence'] ===
+            'number'
+              ? (mainModelResult['unproductive_state_confidence'] as number)
+              : 0,
+          ),
+        );
+
+        if (
+          mainModelResult &&
```

Comment:
I imagine this logic of extracting the result and checking -> reporting can all be unified here and above